Adding dependency walker and test cases

gazunder opened the pull request:
The purpose of this code is to return a list of dependent rules for a given rule. This code takes in a ruleset and a list of rule identifiers to get dependencies for. For each identifier the code will recursively get the the rules that it depends on. Happy to implement any changes/feedback :D 

See #48

plusvic commented on utils/dependency_walker.go:
```diff
@@ -0,0 +1,225 @@
+}
+
+// addNodeChildrenToQue adds a nodes children to the queue
+func addNodeChildrenToQue(queue *[]queueT) {
```
By looking at this function's signature or documentation is not clear what it does. The name implies that the children of some node are added the queue, but which node? By looking at the implementation is clear that its the children of the first node in the queue, but it requires you to read the function's code. 

```
// Adds node's children to queue.
addChildrenToQueue(node ast.Node, queue *[]queueT)
```

plusvic commented on utils/dependency_walker.go:
```diff
@@ -0,0 +1,225 @@
+	"github.com/VirusTotal/gyp/ast"
+)
+
+var conditionKeywords = []string{"int8", "int16", "int32", "uint8", "uint16", "uint32", "int8be", "int16be", "int32be", "uint8be", "uint16be", "uint32be"}
```
This is mainly for identifying built-in functions like `intXX` from other identifiers, so I think it's worth improving the parser by including more information about each identifier in the AST itself. The need for this array comes from a lack of features in the AST. Each `ast.Identifier` struct could contain not only the name, but also the kind of identifier if known. During the parsing we are able to know if an identifier is a rule identifier, or a built-in function, for example.

gazunder commented:
@plusvic thanks for the detailed feedback. I've addressed all of the comments you made in the commits I just added to this PR. 

Some main changes include:
1. `ast.FunctionCall` now has a builtin bool value. This allowed us to remove the `conditionKeywords` from `dependency_walker.go` Added test case for this too.
2. `GetDependencyChainForRules` No now returns a copy of the ruleset, doesn't call `SortRules`, and doesn't mark rules as private
3. Documentation improvements

Let me know if there are any other changes I should make or if I missed anything. Thanks!

gazunder commented on utils/dependency_walker.go:
```diff
@@ -0,0 +1,223 @@
+// GetDependencyChainForRules will recursively find dependencies for a list of rules
+// returnDepAsPrivate will dictate if dependent rules are marked as private
+func GetDependencyChainForRules(ruleset ast.RuleSet, ruleNames ...string) (ast.RuleSet, error) {
+	queue := make(map[string]struct{}) // List of rules to get dependencies for
```
> It is because you don't want to insert duplicate items in the queue?

Yes I don't want to insert duplicate items - I'll rename this to pendingRules.
